### PR TITLE
Add possibility to tag Amazon AWS EBS volumes through an annotation in PVC

### PR DESCRIPTION
**Pull request by scholzj**
**What this PR does / why we need it**:
When running Kubernetes in enterprise environment (such as in my work) it is often required to attach some special tags to some resources such as disks or loadbalancers. Such tags are often used for security (what kind of data is stored on the disk) or cost splitting purposes (which project / department should pay for it). 

On AWS, it is currently possible to tag ELB loadbalancers using the annotation in the service resource (`service.beta.kubernetes.io/aws-load-balancer-additional-resource-tags`). But it is not possible to tag EBS volumes. This PR adds the possibility to tag EBS volumes using annotation from persistent volume claim. The annotation name is used `volume.beta.kubernetes.io/aws-block-storage-additional-resource-tags`.

To avoid unnecessary duplication I also did some minor refactoring for the function parsing tags from annotation so that the same function can be used for both volumes and load balancers. The things are now split between `pkg/volume` and `pkg/cloudprovider`, but it seemed to me better than duplicating code.

**Which issue this PR fixes**
n/a

**Special notes for your reviewer**:
n/a

**Release note**:
```release-note
Add possibility to tag Amazon AWS EBS volumes through an annotation in PVC
```

### CODE REVIEW

File: pkg/cloudprovider/providers/aws/aws.go [hunk 1]
```diff
@@ -3443,3 +3449,31 @@ func recordAwsMetric(actionName string, timeTaken float64, err error) {
...
+				// There is a key and a value, so save it
+				additionalTags[tag[0]] = tag[1]
+			} else if len(tag) == 1 && len(tag[0]) != 0 {
+				// Just "Key"
+				additionalTags[tag[0]] = ""
+			}
+		}
+	}
+
+	return additionalTags
```
Comment (gnufied) [inline pkg/cloudprovider/providers/aws/aws.go]: Can we add some tests for this function?

Comment (scholzj) [inline pkg/cloudprovider/providers/aws/aws.go]: This should be covered by TestGetAdditionalTagsFromAnnotation in pkg/cloudprovider/providers/aws/aws_test.go (this test already existed, I just slightly adapted it to match the other modifications). Do you have some additional tests in mind? I can of course add them.

Comment (gnufied) [inline pkg/cloudprovider/providers/aws/aws.go]: oh that sounds fine then. ignore me comment.

**Comment by gnufied** [comment]
@k8s-bot ok to test

**Comment by gnufied** [comment]
Just a quick update, the idea and feature itself looks sound to me - but in the interest of bringing other cloudproviders to same footing, I was wondering if we should implement this at `pv_controller` like https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/persistentvolume/pv_controller.go#L1312 rather than just for AWS EBS.

I think, it could be useful to other cloudproviders too. cc @jingxu97 @rootfs

**Comment by scholzj** [comment]
@gnufied Sounds reasonable. It looks like implementing this for the volumes accross cloud providers should be quite straight forward. If there is some general agreement, I can rework the code changes (or I can do a separate PR if needed). 

Where would you put the `GetAdditionalTagsFromAnnotation` function so that it can be easily shared between the pv_controller.go and the aws_loadbalancer.go?

**Comment by scholzj** [comment]
BTW: Do the failing `pull-kubernetes-e2e-gce-etcd3` tests have any relation to my changes? It looks to me as if this is an unrelated problem. What is the right approach to get it fixed? Should I just try to restart them until it fixes it self? Thanks.

**Comment by gnufied** [comment]
@scholzj No the failures here are unrelated to your change. They are probably just flakes that we battle every day. :(

**Comment by gnufied** [comment]
cc @kubernetes/sig-storage-pr-reviews

**Comment by scholzj** [comment]
@gnufied Are we still waiting for someone to comment on this? Or should I have a look how to move the vode into pv_controller.go?

**Comment by scholzj** [comment]
Hi @gnufied 

I adapted the PR as you suggested. I moved the tagging of the volumes into the pv_controller so that it works also outside of AWS. The new name of the annotation is `volume.beta.kubernetes.io/additional-tags`. Since I moved the tagging outside of the AWS cloud provider I though that I should also moved the utility function for parsing the tags to a higher level.

On my local host all tests are passing and it also seems to work live in AWS. I haven't tried any other providers.

I would appreciate if you could have a look and let me know if you have any comments. It would be great to get this solved and merged.

Thanks
JAkub

**Comment by gnufied** [comment]
I am yet to review code in detail (I just returned from vacation). @justinsb  does this meets goals of feature request https://github.com/kubernetes/features/issues/300 ? 

I guess at some point we will have to migrate these annotations from beta to GA. but other than that - idea looks pretty sound to me.

File: pkg/cloudprovider/providers/aws/aws_loadbalancer.go [hunk 2]
```diff
@@ -29,39 +29,11 @@ import (
 	"k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/types"
 	"k8s.io/apimachinery/pkg/util/sets"
+	addtionaltags "k8s.io/kubernetes/pkg/util/tags"
```
Comment (justinsb) [inline pkg/cloudprovider/providers/aws/aws_loadbalancer.go]: nit: typo

File: pkg/controller/volume/persistentvolume/pv_controller.go [hunk 3]
```diff
@@ -38,6 +38,7 @@ import (
 	"k8s.io/kubernetes/pkg/controller/volume/events"
 	"k8s.io/kubernetes/pkg/util/goroutinemap"
 	"k8s.io/kubernetes/pkg/util/goroutinemap/exponentialbackoff"
+	addtionaltags "k8s.io/kubernetes/pkg/util/tags"
```
Comment (justinsb) [inline pkg/controller/volume/persistentvolume/pv_controller.go]: typo again

**Comment by justinsb** [comment]
Sorry I missed this earlier.

I actually really like this, and given we do this for ELBs we should do this for EBS volumes also.

We could do this just for AWS without much further discussion IMO.  But if we want to make it cross cloud, that opens up a huge can of worms.  What do we do about clouds that don't support tags, or that have different semantics for allowed keys/values etc.

If this goes back to AWS only I think we can get it in, and then we can discuss some sort of cross-cloud design for 1.9 - sig-cloud should be up and running soon.

/ok-to-test

**Comment by gnufied** [comment]
@justinsb looking at existing pv controller, all cloudprovider's support some sort of tagging, because some tags are added by default from pv controller, this just adds some user provided tags to mix. cc @jsafrane

**Comment by scholzj** [comment]
@justinsb I fixed the typos - thanks for noticing them. 

If needed it should be easy to rollback to the original AWS only version. If it helps we can also use this PR to merge the more straight forward version for AWS only and then I can open a new PR with the version for all cloud providers.

**Comment by scholzj** [comment]
@gnufied @justinsb Hi guys ... anything new about this topic? Thanks.

**Comment by justinsb** [comment]
@kubernetes/sig-gcp-feature-requests @kubernetes/sig-azure-misc    this makes sense on AWS.  But I'm less sure it makes sure on GCP / Azure (or that it should work the same way).  Thoughts?

**Comment by scholzj** [comment]
@justinsb Since there seems to be no feedback - should I just revert this to the original AWS only version so that we can merge this?

**Comment by emmanuel** [comment]
@justinsb @scholzj looks like this 'died in committee'. This is a plea to resurrect the original, limited approach (AWS-only), ship that, and *then* open up and hammer through the discussion about other cloud providers.

**Comment by scholzj** [comment]
@emmanuel I can rebase / update the PR if needed. But merging it is not up to me.

**Comment by pierreozoux** [comment]
@gnufied @justinsb is it possible to reopen this one? Or should we open a new one?

**Comment by almariah** [comment]
Controller and webhook mutations server that could help temporally: https://github.com/almariah/k8s-metadata-injector

**Comment by mlushpenko** [comment]
So is this dead or some work is going on? Same requirement on Azure and in general any cloud provider seems reasonable to me.

**Comment by seb-steuer** [comment]
how is this not a thing yet?

**Comment by cfarrend** [comment]
This would be a really good feature to help us with cost management

**Comment by PierreBeucher** [comment]
It would be great to have this for our needs as well as it seems possible for other resources

**Comment by sshishov** [comment]
Guys, the work is done already, need just rebase and merge. Please have a look!

**Comment by peimanja** [comment]
Any Update on this PR? this is much needed.

**Comment by joel-schaal** [comment]
@josmo: It seems your `/reopen` did not actually reopen this issue. 
@scholzj: maybe you need to do this, since you're the original author ?
Or can someone from the Kubernetes team advise what is the *real* status on that issue ?
Thank you !

**Comment by scholzj** [comment]
@joel-schaal Not sure there is value of re-opening it if nobody reviews / merges this. I'm fine to rebase / update this, but it would be good to first know that it will not be closed again in another 30 days 🤷 .

**Comment by jbury** [comment]
@scholzj can we just ping some folks in for reviews and try to get this in?  There's a lot to read through in the contributing guidelines (I'll take a look later) so I'm not sure if that's taboo, but it'd be nice to guarantee some eyes on this.

**Comment by jbury** [comment]
After looking through a few conflicting contributing related docs, I think our best bet is to just reopen, @ mention the reviewers, wait a week or two and either ping them on slack, or see if we can replace them with alternate relevant owners.  

@scholzj Please re-open this and rebase (or update if needed).  It very much seems like this just got lost after the first round of reviews.  Neither of the required reviewers had blocking comments, so I won't feel too bad about pinging them again for a final readthrough/merge since it should hopefully be very quick for them.  Both reviewers also seem to still be quite active with k8s, so I'd be surprised if we couldn't get ahold of them.

**Comment by joeykhashab** [comment]
Just wanted to add that hopefully this feature gets added, we need to use AWS Backup Service and adding custom tags to the EBS volumes would be great for that.

**Comment by AndrewFarley** [comment]
Ugh, +1?  :P  Can someone make this work, this seems to be 3-years in the desired features list on a variety of issues

**Comment by Lorderot** [comment]
@gnufied @justinsb Is there any issues with this PR? Could we merge it at least for AWS? It's a pity that there's an elegant solution for adding custom tags to ELBs, but none for EBS.

**Comment by AndrewFarley** [comment]
Yeah, this is like a bad joke, 5 years later and its still as frustrating as it was back then.  People can't get proper accounting automatically from services they run in Kubernetes and the resources they use because of the lack of ability to cascade tags into objects.  Can someone please take ownership of this and see this to the finish line?

**Comment by AndrewFarley** [comment]
@wandribeiro That auto-tags EVERY volume in that storage class.  What this PR and feature is about, is allowing per-volume tagging.  If you annotate an PVC, it will tag that SINGLE EBS volume.

To use the above, we would need to (and I have done this) create a single StorageClass for each individual PVC we create, so we can have it tagged appropriately.  This is... crude, at best, and almost offensive that we can't tag still in 2023 per-volume via an annotation

**Comment by ConnorJC3** [comment]
Unfortunately, Kubernetes doesn't provide any volume-specific information like annotations to CSI Drivers except certain "blessed" parameters (volume size, topology, name, etc). Thus, we on the CSI Driver side largely have our hands tied unless/until that happens.

There is an open feature request here to pass annotations to CSI drivers: https://github.com/kubernetes-csi/external-provisioner/issues/86

**Comment by AndrewFarley** [comment]
@ConnorJC3 Although I see this COULD clearly be done at the external-provisioner level, I could also see that you could bypass that entirely if this service was given the RBAC permissions to lookup the annotations which if I'm understanding this pull request, is effectively what this does.

Is there some reason we can't temporarily use something like this Pull Request's code to facilitate this in some temporary way which is noted in the code/comments to eventually replace with native support for CSI annotations being passed through?

I suggest this solution because that external-provisioner issue is 6 years old and has not been solved, the pessimist in me doesn't imagine it will be solved soon.

**Pull request closed without merging**